### PR TITLE
bpo-44010: IDLE: colorize pattern-matching soft keywords

### PR DESCRIPTION
**Pull request by taleinat**
This implements highlighting of the `match`, `case` and `_` [soft keywords](https://docs.python.org/3.10/reference/lexical_analysis.html#soft-keywords), following the syntax defined [in the docs](https://docs.python.org/3.10/reference/compound_stmts.html#the-match-statement) to the extent reasonably possible within the framework of the current colorization method.

https://bugs.python.org/issue44010

### CODE REVIEW

File: Lib/idlelib/colorizer.py
Comment (E-Paine) [inline Lib/idlelib/colorizer.py]: Looking for '\_' as a keyword is a nightmare. There are so many random cases (e.g. in a `sequence_pattern`) and there is no guarantee the start of the statement will be on the same line. Regexes, I believe, are way too limited for this. Maybe we should forget about '\_' for now?

File: Lib/idlelib/colorizer.py
Comment (E-Paine) [inline Lib/idlelib/colorizer.py]: Is this the same as in test_colorizer? If so we can save redundancy and import it from there.

File: Lib/idlelib/colorizer.py
Comment (taleinat) [reply Lib/idlelib/colorizer.py]: This is only specifically for `case _:`, and it seems to work well for that. Treating this specially seems justified as this is common and the only way to define a default `case`.

What other cases are nightmarish? (References would be helpful.)

File: Lib/idlelib/colorizer.py
Comment (taleinat) [reply Lib/idlelib/colorizer.py]: Mostly the same but not entirely. ISTM this bit of duplication for tests is fine.

File: Lib/idlelib/colorizer.py
Comment (E-Paine) [reply Lib/idlelib/colorizer.py]: Yes, it does work well (as far as I can tell) with `case _:`. And I see your point about that being common so maybe we should keep it. One situation I was thinking of was:
```
case [[_],
      _]:
```
Here, both underscores should be highlighted as keywords.

Comment (taleinat) [reply Lib/idlelib/colorizer.py]: Perhaps we could mark any use of `_` in a `case` clause is such a soft-keyword?

File: Lib/idlelib/colorizer.py
Comment (terryjreedy) [inline Lib/idlelib/colorizer.py]:
```suggestion
        # Strings should be entirely colored, including quotes.
        'x', '''x''', "x", \"""x\"""
        'abc\\
        def'
        '''abc\\
        def'''
        # All valid prefixes for unicode and byte strings should be colored.
```
This block and the 5 lines of prefix examples should be moved to the end.

File: Lib/idlelib/colorizer.py
Comment (terryjreedy) [reply Lib/idlelib/colorizer.py]: I believe the main (only?) difference is the addition here of the prefix section.  If this block is kept at the end, which I want, it would be easy to import the common block (inside the htest function to avoid circular import issue), insert it, and then insert prefix block.

I did not check prefixes in automated tests because getting them right is tedious but simple, and easy to check visually, which should be done whenever the file is edited.

File: Lib/idlelib/colorizer.py
Comment (terryjreedy) [inline Lib/idlelib/colorizer.py]: Why the 'PM_' prefix and what is it supposed to mean?  "MATCH", "CASE_NODEFAULT", "CASE_DEFAULT", "UNDERSCORE" are clear and easier to read.

Are the two CASES needed because if it 'case' is matched separately, it will not be available as pre-context for '_'?

File: Lib/idlelib/colorizer.py
Comment (E-Paine) [reply Lib/idlelib/colorizer.py]: I believe that is a reasonable assumption. It seems you cannot use a variable named '\_' in a case statement. For example, the following raises a `SyntaxError` rather than `NameError`:

```
case [_ * 3]:
```

Comment (terryjreedy) [reply Lib/idlelib/colorizer.py]: The match doc in not completely clear to me but I believe it collectively says that isolated '_' is an unbindable keyword in case patterns.  Hence it cannot be used as a bound or binding name -- in patterns.  But this experiment show that it is normal elsewhere, even in match headers, guards, and actions.
```
>>> _ = 'a'
>>> match _:
...     case _ if _: print(_, 'ok')
... 
...     
a ok
```
If we can match any and every isolated '_' in patterns and only in patterns, great.  It does not matter if invalid (as after 'as').  I'm not sure if it is possible.

If we only wanted to match underscore in 'case  _:', we could do it more simply.
```suggestion
        r"(?P<CASE_UNDERSCORE>case[ \t]_)[ \t]*:"
```
Since by default and likely in most custom themes, the keyword background is the same as normal background, including whitespace in the tagged slice would not matter.

**Review by terryjreedy — CHANGES_REQUESTED** [review]
Code review only so far.

File: Lib/idlelib/colorizer.py
Comment (terryjreedy) [reply Lib/idlelib/colorizer.py]: `_ * 3` is invalid *because* `_` is a keyword with no value.  `_` is valid when used properly to mean 'match anything'.
```
>>> match ['a']:
...     case [_]: print('ok')
... 
ok
```
As I tried to say above, misuses that are misuses *because* it is parsed as a keyword should be tagged as a keyword.  In this case, tagging makes it more obvious that the error is because one is attempting to multiply a keyword rather than a value.

**Comment by terryjreedy** [comment]
I now agree that this is much more important than f-string field expressions.  Highlighting helps one read complicated code and also to understand some errors.  Field expressions tend to be short and simple.  Most would get no markup other than 'not string'.

File: Lib/idlelib/colorizer.py
Comment (taleinat) [reply Lib/idlelib/colorizer.py]: `PM` is for "pattern matching", indeed unclear here. Would adding a comment about this suffice in your opinion?

**Comment by mehdi40277** [comment]
at the time of the initial coding, I defined using a formula that had numerical coordinates and each currency code had a unique identity, and at the same time was operating cash under the pseudonym Satoshi Nakamoto. And my goal The induction of banking stability cryptocurrency was created to elevate a revolution in the attitude of the central treasury and the monetary organization, and the idea of ​​creating each sell was in line with a coding program that would create economic transparency in the world and provide financial security to capital owners. Every human being has a new vision and a true belief that the honesty of the thoughts that originate leads the heart to the reasoning of the metaphysical science of the brain. During these several years since the birth of this financial system, many people have abused my position and name and shown themselves in my place. Wow, the growth attitude is to induce the removal of Satoshi Nakamoto and name an identity that, if his name produced an economic revolution, stemmed only from faith and God's direct connection to the intrinsic intellect and intuition that served human society. But I can no longer bear that Such capital that using the sciences of God This kind of plunder is created by opportunists who try to shake this illusion from the darkest parts of everything. The goal is to create the attitude of the birth of superior minds to leave a legacy to the future, as our past left us. In fact, man The further it goes, the closer it actually gets to the past of its predecessors, and this means that we will exist again and again in nature, but with a different appearance and mask that we choose. So the new revolution is the attitude of Bitcoin today towards It starts a new one, and that is creating the business balance of an economic act in relation to certain times, such as the coronary heart disease epidemic in the world, which with this excellence creates the confidence of capital owners in times of such problems and disasters and the existence of Bitcoin financial insurance. The title of Quinn Insurance, which is based on stock exchange transactions, of course, depending on the type of problem, for example, during major illness, the activity to provide medicine, vaccines and items needed by human societies, because the growth of a nation is only when that nation is equal to gender and Race to know that humans based on reason and Judy is judged, not visually. So my friends, brothers, sisters and masters. I once again, as a service to you, proves the motivation of time and not language. So my heart is the place of love and wealth that you have in your heart and You give it to your fellow man with a smile. You are your real wealth. Love is believed to be a seedling in the existence of every hero.

**Comment by mehdi40277** [comment]
at the time of the initial coding, I defined using a formula that had numerical coordinates and each currency code had a unique identity, and at the same time was operating cash under the pseudonym Satoshi Nakamoto. And my goal The induction of banking stability cryptocurrency was created to elevate a revolution in the attitude of the central treasury and the monetary organization, and the idea of ​​creating each sell was in line with a coding program that would create economic transparency in the world and provide financial security to capital owners. Every human being has a new vision and a true belief that the honesty of the thoughts that originate leads the heart to the reasoning of the metaphysical science of the brain. During these several years since the birth of this financial system, many people have abused my position and name and shown themselves in my place. Wow, the growth attitude is to induce the removal of Satoshi Nakamoto and name an identity that, if his name produced an economic revolution, stemmed only from faith and God's direct connection to the intrinsic intellect and intuition that served human society. But I can no longer bear that Such capital that using the sciences of God This kind of plunder is created by opportunists who try to shake this illusion from the darkest parts of everything. The goal is to create the attitude of the birth of superior minds to leave a legacy to the future, as our past left us. In fact, man The further it goes, the closer it actually gets to the past of its predecessors, and this means that we will exist again and again in nature, but with a different appearance and mask that we choose. So the new revolution is the attitude of Bitcoin today towards It starts a new one, and that is creating the business balance of an economic act in relation to certain times, such as the coronary heart disease epidemic in the world, which with this excellence creates the confidence of capital owners in times of such problems and disasters and the existence of Bitcoin financial insurance. The title of Quinn Insurance, which is based on stock exchange transactions, of course, depending on the type of problem, for example, during major illness, the activity to provide medicine, vaccines and items needed by human societies, because the growth of a nation is only when that nation is equal to gender and Race to know that humans based on reason and Judy is judged, not visually. So my friends, brothers, sisters and masters. I once again, as a service to you, proves the motivation of time and not language. So my heart is the place of love and wealth that you have in your heart and You give it to your fellow man with a smile. You are your real wealth. Love is believed to be a seedling in the existence of every hero.

File: Lib/idlelib/colorizer.py
Comment (terryjreedy) [reply Lib/idlelib/colorizer.py]: I strongly prefer deleting it.  To me, it is just noise.

Comment (taleinat) [reply Lib/idlelib/colorizer.py]: Done.

**Comment by taleinat** [comment]
I've refactored the code somewhat and made it mark all lone underscores in `case <...>:` as keyword. I think it'll need more tweak to handle guards, tho.

**Comment by taleinat** [comment]
I still haven't done away with the duplicated test code fragment, but I'll make sure to do so if we move forward with this.

**Comment by terryjreedy** [comment]
'_' is only and always a keyword in patterns, but not in guards or anywhere else.  Simple rule, fortunately.  The guards are standard python, as are the actions and match line. Doc has been patched to say do.  The example from one of the outdated comments 2 day ago:
```
>>> _ = 'a'
>>> match _:
...     case _ if _: print(_, 'ok')
...  
a ok
```

**Comment by E-Paine** [comment]
I am not sure this quite works, sadly. IMO, we need to recursively colour all content in the guard:

```
case _ if ("a" if _ else set()): pass
```

**Comment by taleinat** [comment]
> '_' is only and always a keyword in patterns, but not in guards or anywhere else.

A bare `_` is explicitly, specifically disallowed in pattern-matching "capture groups", i.e. `case <pattern> as _:` is disallowed.

The latest commits *do* mark a bare `_` as a keyword in this case.

**Comment by taleinat** [comment]
> I am not sure this quite works, sadly. IMO, we need to recursively colour all content in the guard:
>
> ```case _ if ("a" if _ else set()): pass```

Fixed, and I've added that example to the tests.

**Comment by taleinat** [comment]
I have made the requested changes; please review again.

File: Lib/idlelib/colorizer.py
Comment (terryjreedy) [inline Lib/idlelib/colorizer.py:33]: When another statement directly follows, I prefer to have the dedent at the end of the statement, instead of before.
```suggestion
        )
```
At the end of a function, I don't care so much.

File: Lib/idlelib/colorizer.py
Comment (terryjreedy) [inline Lib/idlelib/colorizer.py]:
```suggestion
    prog = re.compile("|".join([
                                builtin, comment, string, kw,
                                match_softkw, case_softkw_and_pattern,
                                any("SYNC", [r"\n"])
                              ]),
                      re.DOTALL | re.MULTILINE)
```
Putting arguments on a totally dedented line seems wrong.  In any case, formatting this call with nested calls, one with a list arg,  is awkward.  My suggestion is clearer to me as better indicating which items are part of what.

File: Lib/idlelib/colorizer.py
Comment (terryjreedy) [inline Lib/idlelib/colorizer.py:300]: Coverage says that this condition is never false, hence the else code, `ok = False`, is never executed.

File: Lib/idlelib/colorizer.py
Comment (terryjreedy) [inline Lib/idlelib/colorizer.py:76]: Need at least minimal docstring.

File: Lib/idlelib/colorizer.py
Comment (terryjreedy) [inline Lib/idlelib/colorizer.py:318]: Docstring.

File: Lib/idlelib/colorizer.py
Comment (terryjreedy) [inline Lib/idlelib/colorizer.py]: What is a 'section'?

File: Lib/idlelib/colorizer.py
Comment (terryjreedy) [inline Lib/idlelib/colorizer.py]: Coverage says condition is never false.  It is for something like 'def ++', in which case no highlight is applied.  Numbers are matched even in first position , which I think is a known issue, so 'def 123' wrongly passes as an id. 
'def ++' could be added to one of the lines without changing line numbers.

File: Lib/idlelib/idle_test/test_colorizer.py
Comment (terryjreedy) [inline Lib/idlelib/idle_test/test_colorizer.py:53]: I would like these merged into the examples on 30+ *without* changing line numbers.  For instance, change an 'x' to 'match x'.  I am willing to do it, maybe after this is merged.  See comment above about 'def non-name'.

File: Lib/idlelib/idle_test/test_colorizer.py
Comment (terryjreedy) [inline Lib/idlelib/idle_test/test_colorizer.py]: Now 99%.  Try to fix on one of followups.  I am not going to rerun tests for this.

**Review by terryjreedy — COMMENTED** [review]
I think that this is nearly ready.  I made a couple of comments each about formats, docstrings, and coverage.

The docstrings are the most important for me or anyone other than Tal to be able respond to bug reports.  I will read the new tagging code tomorrow to see if I find anything that needs explanation or comment.

I am nervous about major change to the main function, but nearly complete coverage and what seems like testing is a major confidence booster.  I would like to see perhaps a page of example code attached to the issue that we can manually check.  All the examples in the doc would be a start.  Do match tests have code snippets?  (And where would  they be?)

Speed matters, and we seem to be good.  Code is recolorized from some sync point (beginning of line, further along?) with each keystroke.  Bulk recolorizing the 2300 lines of configdialog.py (with line numbers) after a font size change takes about a second for me.  (Jump to the bottom with ^End, open dialog, change size, click OK.)   Without a stopwatch, before** and after seem about the same.

File: Lib/idlelib/colorizer.py
Comment (taleinat) [reply Lib/idlelib/colorizer.py:33]: Done.

File: Lib/idlelib/colorizer.py
Comment (taleinat) [reply Lib/idlelib/colorizer.py]: Done.

File: Lib/idlelib/colorizer.py
Comment (taleinat) [reply Lib/idlelib/colorizer.py:300]: I've added a test case that should exercise this code path.

File: Lib/idlelib/colorizer.py
Comment (taleinat) [reply Lib/idlelib/colorizer.py:76]: Done.

File: Lib/idlelib/colorizer.py
Comment (taleinat) [reply Lib/idlelib/colorizer.py:318]: Done.

File: Lib/idlelib/colorizer.py
Comment (taleinat) [reply Lib/idlelib/colorizer.py]: I've reworded this doc-string.

File: Lib/idlelib/colorizer.py
Comment (taleinat) [reply Lib/idlelib/colorizer.py]: I've added several test cases that exercise this code path.

BTW, fixing the incorrect highlighting of `123` in `def 123` is simple. Would you like that done here or in a separate PR?

File: Lib/idlelib/idle_test/test_colorizer.py
Comment (taleinat) [reply Lib/idlelib/idle_test/test_colorizer.py:53]: Argh, that would require changing a lot of indexes back to other values... Can't we keep the new values which I've painstakingly verified to be correct?

File: Lib/idlelib/idle_test/test_colorizer.py
Comment (taleinat) [reply Lib/idlelib/idle_test/test_colorizer.py:1]: Soon to be 100%!

File: Lib/idlelib/idle_test/test_colorizer.py
Comment (taleinat) [reply Lib/idlelib/idle_test/test_colorizer.py:53]: I have a branch ready for a followup PR that will make this entirely moot, by using source text with clear annotations of expected highlighting tag ranges, instead of the current method which is very hard to read and maintain.

**Comment by taleinat** [comment]
I have made the requested changes; please review again.

File: Lib/idlelib/colorizer.py
Comment (taleinat) [reply Lib/idlelib/colorizer.py]: > Numbers are matched even in first position , which I think is a known issue, so 'def 123' wrongly passes as an id.

Is there an existing bpo for this?

Comment (terryjreedy) [reply Lib/idlelib/colorizer.py]: No. And yes, let's defer to followup.

File: Lib/idlelib/idle_test/test_colorizer.py
Comment (terryjreedy) [reply Lib/idlelib/idle_test/test_colorizer.py:53]: Fine. Defer to followup.

**Comment by terryjreedy** [comment]
Should I merge if satified after review?

File: Lib/idlelib/colorizer.py
Comment (terryjreedy) [reply Lib/idlelib/colorizer.py]: Two sources have been combined into one.

File: Lib/idlelib/colorizer.py
Comment (terryjreedy) [reply Lib/idlelib/colorizer.py:300]: verified

File: Lib/idlelib/colorizer.py
Comment (terryjreedy) [reply Lib/idlelib/colorizer.py]: Coverage verified.

File: Lib/idlelib/colorizer.py
Comment (terryjreedy) [inline Lib/idlelib/colorizer.py:316]: Coverage says the conditional above is execution both ways, the the return is not.  This is the only thing short of 100%. I am going to assume that this is a glitch in my older version of coverage.

**Comment by terryjreedy** [comment]
Thank you for the updates.  Looks good, so I am merging and we can move onward.

For this file, the htest serves as a nice and nearly complete integration test.  The only manual action I added was to copy paste the match statement into Shell to make sure no bizarre interaction with sidebar or something.  There isn't.

To do: name number patch and test paste.  Update coverage % with latter.

I need to update coverage version and idlelib.News.

**Comment by taleinat** [comment]
Yes, it's good that you caught that... The colorization is currently inconsistent while editing `match` blocks :(

That's inherent to how the colorizer currently works and I assume you wouldn't want us to make any major changes to that right now.

Let me see if I can think of something simpler and more consistent that will be good enough for now...

**Comment by taleinat** [comment]
To clarify (and think aloud a bit), the root of the problem is that unlike triple-quoted strings, it's difficult for the colorizer to recognize when a `case` pattern starts vs. other possible uses of a `case` token, since that requires context which the colorizer doesn't have, since it isn't using a full parser.

**Comment by taleinat** [comment]
Please check the latest commit. It doesn't handle some of the previous cases, but it works more consistently, and seems "good enough" to me.

**Comment by terryjreedy** [comment]
A truly complete integration test would have to test color changes as text is entered and deleted.  As i, n, t, e are entered, color goes from normal to keyword to builtin to normal, and back again as chars are deleted.  With the soft keywords, there are two options: a) color immediately and uncolor when clearly not a keyword; b) leave normal until sure is a keyword.  

Before and after last patch, initial 'match' is colored immediately.  Uncolored if followed by small number of listed symbols, starting with '='.  This list should include all binary-only operators as any of them forces match to be identifier.  Also, all keywords should uncolor 'match'.  Certainly, 'match is a' and 'match in a' are legal expressions in which match is an identifier.  

Before last patch, coloring initial 'case' was delayed.  Now it is not.  I think it better that initial 'match' and 'case' are handled the same.  I think 'mark' as default is sensible since most initial uses of those will be as keywords.

If 'case _:', _ is marked.  In 'case _ if x:' it no longer is, but I think it should be.  It seem to me that 'if' should and can be treated as a pattern terminator the same as ':'.

IDLE's doc has a subsection on "Text colors".  We need to add a sentence about soft keywords and can say that with colorizer's current design, perfection is impossible.

**Comment by taleinat** [comment]
I have made the requested changes; please review again.

**Comment by terryjreedy** [comment]
The version numbers and other extraneous material in help.html are outside of the part of the file displayed, which is  <div class='section>, lines 103 to 878.   They have always been the same in all active versions so that there is never an issue with backporting.

This is the first time we have not backported a doc change to all versions.  So if we do nothing, there would be  merge conflicts the next time we backport a doc change to 3.9.  To avoid this, help.html should be copied, after merging, to the 3.9 branch.  The new paragraph, lines 584 to 588, should be deleted.  The change on 693, which should have been incorporated into the file previously, and all other changes, should be merged into 3.9.

I have considered moving the 'section' extraction mechanism from display to copying.  The result would have to be tested and  might be less robust against sphinx formatting changes, but the diffs and the code would be simpler.

**Comment by taleinat** [comment]
@terryjreedy, do you want anything else done here? Are you okay with this being merged in its current form?

> To avoid this, help.html should be copied, after merging, to the 3.9 branch. The new paragraph, lines 584 to 588, should be deleted. The change on 693, which should have been incorporated into the file previously, and all other changes, should be merged into 3.9.

Alright, we can take care of that in a followup PR.

**Comment by terryjreedy** [comment]
News entries look fine, so go ahead.  I will handle idlelib/NEWS separately.  Also 3.9 doc patch if you don't.

**Comment by taleinat** [comment]
> Alright, we can take care of that in a followup PR.

See PR GH-26239.

**Pull request merged**
